File: tests/steps/4/step4.cpp
#include <QueryStepsPostConfig.h>
#include <importCORBA.h>
#include <importUAKGQuery.h>
#include <UAKGQuery2Init.h>

#include <stdlib.h>
#include <../3/printLib.h>
 //for atoi

/*
 $Id: step4.cpp,v 1.23 2002-12-04 09:29:31 srkr Exp $
*/

#include <iostream>
using namespace std;
using namespace UAKGQuery2;

CORBA::ORB_ptr myORB;


RecordSet_init* recordSetFactory;


    


void run_query(QueryManager_ptr queryManager);

int main(int argc, char** argv)
{
  try {

#ifdef CORBA_ORB_INIT_HAVE_3_ARGS
    CORBA::ORB_var orb = CORBA::ORB_init(argc, argv, CORBA_ORB_INIT_THIRD_ARG);
#else
    CORBA::ORB_var orb = CORBA::ORB_init(argc, argv);
#endif
    initUAKGQuery2(orb.in());

    CORBA::ValueFactory vf=orb->lookup_value_factory(
                             "IDL:gradsoft.kiev.ua/UAKGQuery2/RecordSet:1.0");
    recordSetFactory=RecordSet_init::_downcast(vf);
    if (recordSetFactory==NULL) {
        cerr <<  "Internal error: can't find valuetype factory for RecordSet" 
             << endl;
        return 1;
    }


    CORBA::Object_var obj;
    try {
       obj = orb->resolve_initial_references("UAKGQueryService");
    }catch(const CORBA::ORB::InvalidName&){
       cerr << argv[0] << ": can't resolve UAKGQueryService" << endl;
       return 1;
    }

    if(CORBA::is_nil(obj)) {
      cerr << argv[0] << ": UAKGQueryService is a nil object reference" << endl;
      return 1;
    }

    DBConnectionManager_var dbManager = DBConnectionManager::_narrow(obj);
    if (CORBA::is_nil(dbManager)) {
      cerr << argv[0] << ": can't narrow dbManager to correct type" << endl;
      return 1;
    } 
    QueryManager_var queryManager;
    try {
      queryManager = 
         dbManager->create_query_manager("intbill","intbill","","Oracle8","");
    }catch(QueryManagerNotFound){
      cerr << argv[0] <<": can't find query manager." << endl;
      return 1;
    }
    //*****
    // added:
    run_query(queryManager);

    //time to disconnect.
    //
    cout << "disconnecting." << endl;
    queryManager->destroy();
    orb->destroy();
  }catch(const CORBA::SystemException& ex){
    cerr << ex << endl;
    return 1;
  }
  return 0;
}

void run_query(QueryManager_ptr queryManager)
{
 static const char* query_text = "SELECT id FROM parties WHERE id = :id";

 RecordSet_var recordSet = recordSetFactory->create();

 recordSet->addColumn(":id",TypeLong);

 try {
   Query_var query = queryManager->create_query(query_text,"");
   int quit=0;
   query->prepare_query(recordSet);
   recordSet->addRow();
   while(!quit)
   {
     cout << "enter empno or 'q' to quit:" ;
     cout.flush();
     static char tname[12]; 
     cin.getline(tname,12,'\n');
     if (tname[0]=='\0') continue; 
     if (strcmp(tname,"q")==0) break;
     cout << "1" << endl;
     recordSet->setLongAt(0,0,atoi(tname));
     cout << "2" << endl;
     query->execute(recordSet);
     cout << "3" << endl;
     CORBA::Boolean more=1;
     cout << "4" << endl;
     RecordSet_var rs = query->fetch(0,more);
     printRecordSet(cout,rs);
     cout << "5" << endl;
   }
   query->destroy();
 }catch( const InvalidFieldType&  ex){
   cerr << "InvalidFieldType" << endl;
 }catch(const QueryFlagInvalid& ex){
   cerr << "QueryFlagInvalid:" << ex.why << endl;
 }catch(const QueryProcessingError& ex){
   cerr << "QueryProcessingError" << endl;
   cerr << ex.why << endl;
 }
}

File: tests/steps/4/step4.java
import java.io.*;
import java.util.*;

import org.omg.CORBA.*;
import org.omg.CosQuery.*;
import org.omg.CORBA.ORBPackage.*;
import org.omg.CosQueryCollection.*;

import ua.kiev.grad.RC.*;
import ua.kiev.grad.CosQuery.*;
import ua.kiev.grad.CosQuery.RecordDescriptionHolder;
import ua.kiev.grad.CosQuery.RecordSeqHolder;
import ua.kiev.grad.CosQuery.FieldDescription;
import ua.kiev.grad.CORBAView.*;    

public class step4
{
    private static ORB orb;
    public static void main(String args[])
    {
        try
        {
            Properties props = System.getProperties();
            props.put("org.omg.CORBA.ORBClass", "com.ooc.CORBA.ORB");
            props.put("org.omg.CORBA.ORBSingletonClass", "com.ooc.CORBA.ORBSingleton");
            System.setProperties(props);
            ORB orb = ORB.init(args, props);
            org.omg.CORBA.Object obj = null;
            DBConnectionManager dbcm = null;
            try
            {
                obj = orb.resolve_initial_references("UAKGQueryService");
            }
            catch(InvalidName ex)
            {
                System.out.println("UAKGQueryService service is not available");
                return;
            }
            if(obj == null)
            {
                System.out.println("UAKGQueryService object reference is invalid");
                return;
            }
            dbcm = DBConnectionManagerHelper.narrow(obj);
            if(dbcm == null)
            {
                System.out.println("Can't narrow DBConnection to correct type");
                return;
            }                                              

            UAKGQueryManager queryManager = null;
            try 
            {
                queryManager = dbcm.createQueryManager("scott","tiger","sun_db","Oracle8","");
            }
            catch(QueryManagerNotFound ex)
            {
                System.out.println("Can't find query manager.");
                return;
            }
            printLib.orb = orb;
            run_query(queryManager);
            System.out.println("disconnecting.");
        }
        catch(SystemException ex)
        {
            System.out.println("SystemException: " + ex);
            return;
        }
        System.out.println("Success!!!");
    }
    private static void run_query(UAKGQueryManager queryManager)
    {
        String query_text = "SELECT * FROM tab WHERE tname LIKE :tname";
        ua.kiev.grad.CosQuery.RecordDescriptionHolder s1RecordDescriptionHolder = new ua.kiev.grad.CosQuery.RecordDescriptionHolder();
        ua.kiev.grad.CosQuery.FieldDescription fd = new ua.kiev.grad.CosQuery.FieldDescription(":tname", ValueType.TypeString, 30, (short)0, (short)0);
        s1RecordDescriptionHolder.value = new ua.kiev.grad.CosQuery.FieldDescription[1];
        s1RecordDescriptionHolder.value[0] = fd;
        RecordSeqHolder s1RecordSeqHolder = new RecordSeqHolder();
        s1RecordSeqHolder.value = new org.omg.CosQueryCollection.FieldValue[1][1];
        s1RecordSeqHolder.value[0][0] = new org.omg.CosQueryCollection.FieldValue();
        try 
        {
            UAKGQuery query = queryManager.create_query(query_text, 0);
            query.prepare_query(s1RecordDescriptionHolder.value);
            while(true)
            {
                System.out.print("enter tname or 'q' to quit:");
                byte[] arb_tname = new byte[30]; 
                System.in.read(arb_tname);
                if ((char)arb_tname[0] == '\r') continue; 
                if ((char)arb_tname[0] == 'q' | (char)arb_tname[0] == 'Q') break;
                String str_tname = new String(arb_tname);
                str_tname = str_tname.substring(0, str_tname.indexOf("\n") - 1);
                FieldValueAccess.setString(s1RecordSeqHolder.value[0][0], str_tname);
                query.execute_records(s1RecordSeqHolder.value);
                BooleanHolder more = new BooleanHolder(true);
                RecordSeqHolder recordSeqHolder = new RecordSeqHolder();
                recordSeqHolder.value = query.fetch_records(1, more);
                printLib.printRecordSeq(System.out, recordSeqHolder);
            }
            query.destroy();
        }
        catch(IOException ex)
        {
            System.out.println("IOException: " + ex);
        }
        catch(CORBAViewException ex)
        {
            System.out.println("CORBAViewException: " + ex);
        }
        catch(QueryTypeInvalid ex)
        {
            System.out.println("QueryTypeInvalid");
        }
        catch(QueryInvalid ex)
        {
            System.out.println("QueryInvalid");
        }
        catch(QueryProcessingError ex)
        {
            System.out.println("QueryProcessingError");
            System.out.println(ex.why);
        }
    }
}